cost.go

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package celvet
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/google/cel-go/cel"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation"
+	structuralschema "k8s.io/apiextensions-apiserver/pkg/apiserver/schema"
+	schemacel "k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel"
+)
+
+type costError struct {
+	Program cel.Program
+	Name    string
+	Cost    uint64
+}
+
+func (c *costError) Error() string {
+	exceedFactor := float64(c.Cost) / float64(validation.StaticEstimatedCostLimit)
+	return fmt.Sprintf("expression at %q exceeded budget by factor of %.1fx", c.Name, exceedFactor)
+}
+
+func CheckExprCost(schema *structuralschema.Structural) []*costError {
+	return checkExprCost(schema, "<root>", rootCostInfo())
+}
+
+func checkExprCost(schema *structuralschema.Structural, name string, nodeCostInfo costInfo) []*costError {
+	results, err := schemacel.Compile(schema, false, schemacel.PerCallLimit)
+	if err != nil {
+		// TODO(DangerOnTheRanger): what's the proper way to handle compilation errors here?
```

Comment:
What about to treat compilation err as a level up err separately? I would prefer to have all possible err returned to user as once if possible. And the compilation err for one expression shouldn't affect the check for following expressions.

Reply:
Makes sense - I'll mark this as resolved once I add that change.